### ISSUE TITLE
Parse and warn about merge conflict markers

### ISSUE THREAD

**Issue by j-f1** [eslint/espree#350]
The TypeScript parser can parse merge conflict markers, allowing it to provide a warning message to the user. It then takes the first branch of the conflict, as shown in [this AST Explorer example](https://astexplorer.net/#/gist/8e4db4a300841855952e3b807f37ae04/faf71076a8f771726a1dec2df3d9b50e431abfbc).

Can espree parse these and fail (or warn) with an informative message?

Here’s [the implementation in TypeScript](https://github.com/Microsoft/TypeScript/blob/f667357aad0ef7b9f8d3f4ee10f43add953521bf/src/compiler/scanner.ts#L534-L588) if you’d like to reference it.

**Comment by not-an-aardvark** [comment]
Thanks for the proposal. I don't really feel strongly about this, but I'm not sure this is a good idea because we would be returning a valid AST for something that's not valid JavaScript code. That could have the effect of confusing downstream tools, because e.g. eslint rules would notice a bunch of unexpected tokens in the file as a result of merge conflict markers.

**Comment by nzakas** [comment]
This isn't appropriate for Espree. Most merge conflicts will already throw a parsing error, so I think this is already handled appropriately.

**Comment by j-f1** [comment]
@nzakas @not-an-aardvark 

> I'm not sure this is a good idea because we would be returning a valid AST for something that's not valid JavaScript code

Why not have something like this?

```js
someCodeHere()

>>>>>>> deadbeef
// ^ SyntaxError: Unexpected merge conflict marker
moreCodeHere()
=======
oopsIChangedThis()
<<<<<<< HEAD

evenMoreCodeHere()
```

**Comment by platinumazure** [comment]
I think this could be accomplished by writing a parser which wraps espree and also comes with some sort of git/patch/diff library that can identify if a file has diff/patch markers.

Code could look something like this:

```js
const espree = require("espree");
const isDiff = require("some-diff-patch-library").isDiff;

exports.parse = function (sourceText, options) {
    if (isDiff(sourceText)) {
        // Find offending lines, throw a parse error
    } else {
        return espree.parse(sourceText, options);
    }
};
```

**Comment by nzakas** [comment]
@j-f1 What you're proposing is a lot of work for a very small benefit in an edge case. We don't produce custom error messages for other parsing errors, so I don't think we should be doing it here either. The line and column number returned by the parser tell you where to look, and I think that's about as far as we should go.
